### PR TITLE
ast: mark ast.Num etc. as deprecated

### PR DESCRIPTION
**Pull request by JelleZijlstra**
PEP 702 was just accepted. As the decorator should simply be ignored
by type checkers that do not support it yet, I don't see anything
blocking us from using it now.

This is the first simple-looking deprecation in
https://docs.python.org/3.12/whatsnew/3.12.html#deprecated
I'll see what CI has to say here before looking at others.

### CODE REVIEW

**Review by srittau — APPROVED** [review]

**Review by AlexWaygood — APPROVED** [review]
Nice, those are very useful error messages emitted by pyright!

File: stdlib/ast.pyi [hunk 1]
```diff
@@ -4,27 +4,32 @@ from _ast import *
...
+from typing_extensions import Literal, deprecated
 
 if sys.version_info >= (3, 8):
     class _ABC(type):
         if sys.version_info >= (3, 9):
             def __init__(cls, *args: Unused) -> None: ...
 
+    # TODO: flake8-pyi should allow long strings here
+    # https://github.com/PyCQA/flake8-pyi/issues/443
+    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")  # noqa: Y053
```
Comment (AlexWaygood) [inline stdlib/ast.pyi:16]:
```suggestion
    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")
```

File: stdlib/ast.pyi [hunk 2]
```diff
@@ -4,27 +4,32 @@ from _ast import *
...
     class Num(Constant, metaclass=_ABC):
         value: int | float | complex
-
+    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")  # noqa: Y053
```
Comment (AlexWaygood) [inline stdlib/ast.pyi:19]:
```suggestion
    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")
```

File: stdlib/ast.pyi [hunk 3]
```diff
@@ -4,27 +4,32 @@ from _ast import *
...
     class Str(Constant, metaclass=_ABC):
         value: str
         # Aliases for value, for backwards compatibility
         s: str
-
+    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")  # noqa: Y053
```
Comment (AlexWaygood) [inline stdlib/ast.pyi:24]:
```suggestion
    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")
```

File: stdlib/ast.pyi [hunk 4]
```diff
@@ -4,27 +4,32 @@ from _ast import *
...
     class Bytes(Constant, metaclass=_ABC):
         value: bytes
         # Aliases for value, for backwards compatibility
         s: bytes
-
+    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")  # noqa: Y053
```
Comment (AlexWaygood) [inline stdlib/ast.pyi:29]:
```suggestion
    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")
```

File: stdlib/ast.pyi [hunk 5]
```diff
@@ -4,27 +4,32 @@ from _ast import *
...
     class NameConstant(Constant, metaclass=_ABC): ...
+
+    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")  # noqa: Y053
```
Comment (AlexWaygood) [inline stdlib/ast.pyi:32]:
```suggestion
    @deprecated("Replaced by ast.Constant; removal scheduled for Python 3.14")
```

**Pull request merged**